### PR TITLE
[apiv2] Implement crawlerAddDoc, crawlerCommit, crawlerCommitInfo functions

### LINKED ISSUE #404: Change API for crawler
The goal of this story is to define a new API suitable to implement change crawler. It will be used for GitLab. The goal is to define a updateable entities by oldest order: *organization* and *project*:

For each entity (or system api), the api let you query the oldest element, for example:

Base api:

- Monocle.Backend.get_last_updated :: entity_type -> m [entity]
- Monocle.Backend.updated_last_updated_at :: entity_type -> entity -> utctime -> m ()

For the *organzation* entity:

- Lentille.fetch_project :: organization -> m [project]
- Monocle.Backend.add_project :: organization -> [project] -> m ()

```haskell
projectCrawler = do
  lasted_updated_at <- getCurrentTime
  oldest_org <- Monocle.Backend.get_last_updated "organization"

  projects <- Lentille.fetch_project oldest_org
  Monocle.Backend.add_project oldest_org projects

  Monocle.Backend.updated_last_updated_at "organization" oldest_org latest_udpated_at
```

For the *project* entity:

- Lentille.fetch_pr :: project -> m [pr]
- Lentille.fetch_pr_commits :: change -> m [commits]
- Monocle.Backend.change_add :: [documents] -> m ()

```haskell
prCrawler = do
  lasted_updated_at <- getCurrentTime
  oldest_project <- Monocle.Backend.get_last_updated "project"

  pr <- Lentille.fetch_pr oldest_project
  -- stop when pr age > oldest_project age
  commits <- Monocle.Backend.change_add pr
  let changes = toDocuments pr commits
  Monocle.Backend.change_add changes

  Monocle.Backend.updated_last_updated_at "project" oldest_project latest_udpated_at
```

### PR DESCRIPTION
**Pull request by morucci**
resolves #404

### CODE REVIEW

File: haskell/src/Monocle/Api/Server.hs [hunk 1]
```diff
@@ -29,14 +30,101 @@ configHealth = const $ pure response
...
+apiKeyMatch indexName crawlerName apiKey = True
+
 crawlerAddDoc :: CrawlerPB.AddDocRequest -> AppM CrawlerPB.AddDocResponse
-crawlerAddDoc = undefined
+crawlerAddDoc request = do
+  Env {bhEnv = bhEnv} <- ask
+  let (CrawlerPB.AddDocRequest indexName crawlerName apiKey entity changes events) = request
+  case (indexExist indexName, crawlerExist indexName crawlerName, apiKeyMatch indexName crawlerName apiKey) of
+    (True, True, True) -> do
+      case toEntity entity of
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Api/Server.hs:46]: It seems like these checks can be rewritten using this trick: https://www.haskellforall.com/2021/05/the-trick-to-avoid-deeply-nested-error.html

In this case, that would look like something like this:

```haskell
orStop :: Maybe a -> b -> Either b a
orStop = undefined

getResp :: IO CrawlerPB.AddDocResponse
getResp = either (pure . toErrorResponse) go validateRequest
  where
    validateRequest :: Either CrawlerPB.AddDocError ()
    validateRequest = do
      index <- lookupIndex inderName `orStop` AddDocErrorAddUnknownIndex
      crawler <- lookupCrawler index crawlerName `orStop` AddDocErrorAddUnknownCrawler
      when (crawlerKey /= apiKey) (Left AddDocErrorAddUnknownApiKey)
      pure ()
    -- the validation doesn't produce anything, so we `_` the result here
    go _ = do
      I.indexChanges ...
```

**Review by TristanCacqueray — COMMENTED** [review]
LGTM

File: haskell/src/Monocle/Api/Server.hs (hunk 1 above)
Comment (TristanCacqueray) [reply haskell/src/Monocle/Api/Server.hs:46]: Here is an example for the search query validation: https://github.com/change-metrics/monocle/pull/426/commits/a952c4f925794610e1eaa8e32df05d5c01c34d73 . Notice that when all the validation happens in the Either, then we don't need to combine with `orStop`.

Though that is really cosmetic, your change LGTM, although the mock need to be removed :-)

**Comment by TristanCacqueray** [comment]
> _No description provided._

The PR description could mention `Fixes #404`

File: haskell/src/Monocle/Api/Server.hs (hunk 1 above)
Comment (morucci) [reply haskell/src/Monocle/Api/Server.hs]: Thanks for the Either trick, It not strength forward for me as of now. I've tried but it's taking more time time than expected and I figured out I need to tackle function like lookupCrawler first.

File: haskell/src/Monocle/Api/Server.hs [hunk 2]
```diff
@@ -30,14 +31,99 @@ configHealth = const $ pure response
   where
     response = ConfigPB.HealthResponse "api running"
 
+-- Mock to bind to function from another PR
+crawlerExist indexName crawlerName = True
+
+apiKeyMatch indexName crawlerName apiKey = True
+
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Api/Server.hs:38]:
```suggestion
```

**Review by TristanCacqueray — APPROVED** [review]

**Pull request merged**